### PR TITLE
Add terraform deployment as an example.

### PR DESCRIPTION
**Pull request by itsdalmo**
This PR adds a terraform template for deploying lifecycled to an autoscaling group with the lifecycle hook set up (see the included `README.md`). The template should upload `lifecycled-linux-amd64` from the `build/` folder after cloning the repository and running `make release`, so it can be used to debug new changes.

The `cloud-config.yml` should be discussed, but for now I've done the following:
- Included `log-file=/var/log/lifecycled.log` as a parameter (makes this PR dependent on #32).
- The systemd unit for `lifecycled.service` differs from the one included in this repo in several ways:
  - `After=network-online.target`: specifies that `lifecycled` should only be started after the network is online (and shut down before the network). 
  - `Restart=on-failure`: Restart `lifecycled` if it exits with an error. I chose this over `always` because (in the future) it allows us to have a "happy path" which shuts down `lifecycled` immediately after completing the lifecycle hook, instead of waiting for a `SIGTERM`. In this case we don't want it to restart again.
  - `RestartSec=1m`: Since the queue should always be cleaned on exit, it cannot be recreated within 1min of being deleted, so there is no point in restarting before the 1min is up.
  - `TimeoutStopSec=5m`: Added because at least on Amazon Linux 2 I experienced that manual interrupts (`systemctl stop lifecycled.service`) would immediately log that the `SIGTERM` had timed out and issue a `SIGKILL`. This happened even though the default timeout should be 90 seconds.

Note that I'm not a `systemd` expert my any means, so I think the `cloud-config.yml` warrants careful inspection and/or discussion 😄 Also, this should be tested some more once #32 is merge!

### CODE REVIEW

**Comment by itsdalmo** [comment]
Set a WIP flag until we've had a chance to discuss and test this more extensively 👍

**Comment by lox** [comment]
Hah, I wrote almost exactly the same thing in bash and cloudformation in https://github.com/buildkite/lifecycled/pull/14.

**Comment by lox** [comment]
Interesting, those changes to the systemd file sound like they might be generally sensible?

**Comment by lox** [comment]
Happy to merge this when you're done @itsdalmo.

File: terraform/example.tf [hunk 1]
```diff
@@ -0,0 +1,63 @@
+terraform {
+  required_version = "0.11.8"
+}
+
+provider "aws" {
+  version = "1.34.0"
+  region  = "eu-west-1"
```
Comment (lox) [inline terraform/example.tf]: Is it possible to not hardcode this?

**Comment by itsdalmo** [comment]
Good catch on the hardcoded region restriction. I've fixed that and the parts that made this PR reliant on #32 (now closed). We'll need to add back some flags when #38 is merged. I've tested it and everything works as expected (`lifecycled` starts, and has the necessary permissions to do it's job). I.e. it's good to merge 👍

**Comment by lox** [comment]
I got this error:

```
Error: Error applying plan:

1 error(s) occurred:

* module.example.module.asg.aws_launch_configuration.main: 1 error(s) occurred:

* aws_launch_configuration.main: Error creating launch configuration: ValidationError: The key pair 'lifecycled-example' does not exist
	status code: 400, request id: dd75de48-b166-11e8-8877-654a44875dc0

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by itsdalmo** [comment]
That error means you need to create an EC2 Key Pair named `lifecycled-example`. Or you can change that value in `example.tf` to a key pair which already exists, and for which you have access to the `.pem` file (so you can SSH into the instance).

Could add this info to README?

**Pull request merged**